boot/bootutil/src/swap_move.c

```diff
@@ -0,0 +1,506 @@
+    if (bs->state == BOOT_STATUS_STATE_1) {
+        rc = boot_erase_region(fap_primary_slot, pri_off, sz);
+        assert(rc == 0);
+
+        rc = boot_copy_region(state, fap_secondary_slot, fap_primary_slot,
+                              sec_up_off, pri_off, sz);
+        assert(rc == 0);
+
+        rc = boot_write_status(state, bs);
+        bs->idx++;
+        bs->state = BOOT_STATUS_STATE_0;
+        BOOT_STATUS_ASSERT(rc == 0);
+    }
+
+    flash_area_close(fap_primary_slot);
+    flash_area_close(fap_secondary_slot);
+}
+
+/*
+ * While starting a new swap the primary slot is initialized from the data
+ * in the secondary slot. If this is detected as a revert, no information in
+ * the secondary slot exists; so, if the swap is interrupted at the "right"
+ * time the swap will not revert. This function handles the issue by faking
+ * a permanent upgrade in case a reset occurs.
+ */
+void
+fixup_revert(const struct boot_loader_state *state, struct boot_status *bs)
+{
+    struct boot_swap_state swap_state;
+    const struct flash_area *fap;
+    int rc;
+    uint8_t image_index;
+
+#if (BOOT_IMAGE_NUMBER == 1)
+    (void)state;
+#endif
+
+    /* No fixup required */
+    if (bs->swap_type != BOOT_SWAP_TYPE_REVERT ||
+        bs->op != BOOT_STATUS_OP_MOVE ||
+        bs->idx != BOOT_STATUS_IDX_0) {
+        return;
+    }
+
+    image_index = BOOT_CURR_IMG(state);
+
+    rc = boot_read_swap_state_by_id(FLASH_AREA_IMAGE_SECONDARY(image_index),
+            &swap_state);
+    assert(rc == 0);
+
+    BOOT_LOG_SWAP_STATE("Secondary image", &swap_state);
+
+    if (swap_state.magic == BOOT_MAGIC_UNSET) {
+        rc = flash_area_open(FLASH_AREA_IMAGE_SECONDARY(image_index), &fap);
+        assert (rc == 0);
+
+        rc = boot_erase_trailer_sectors(state, fap);
+        assert(rc == 0);
+
+        rc = boot_write_image_ok(fap);
+        assert(rc == 0);
+
+        rc = boot_write_swap_size(fap, bs->swap_size);
+        assert(rc == 0);
+
+#ifdef MCUBOOT_ENC_IMAGES
+        rc = boot_write_enc_key(fap, 0, bs->enckey[0]);
+        assert(rc == 0);
+
+        rc = boot_write_enc_key(fap, 1, bs->enckey[1]);
+        assert(rc == 0);
+#endif
+
+        rc = boot_write_magic(fap);
+        assert(rc == 0);
+
+        flash_area_close(fap);
+    }
+}
+
+void
+boot_perform_swap(struct boot_loader_state *state, struct boot_status *bs,
+        uint32_t copy_size)
+{
+    uint32_t sz;
+    uint32_t idx;
+    uint32_t slot_size;
+
+    slot_size = 0;
+    g_last_idx = 0;
+
+    //FIXME: assume all sectors of same size and just do math here...
+    sz = boot_img_sector_size(state, BOOT_SECONDARY_SLOT, 0);
+    while (1) {
+        slot_size += sz;
+        /* Skip to next sector because all sectors will be moved up. */
+        g_last_idx++;
+        if (slot_size >= copy_size) {
+            break;
+        }
+    }
+
+    fixup_revert(state, bs);
+
+    /* FIXME: assure last sector does not overrun trailer */
+
+    if (bs->op == BOOT_STATUS_OP_MOVE) {
```

Comment:
honestly: I would exclude flash_areas opening/closing and closing from `boot_move_sector_up()` and 'boot_swap_sectors()' and put it the opening before this `if` and the closing after the 'while' with `boot_swap_sectors` inside.
This would simplifies code of these sub-function a little, reduces footprint, speed up execution a little...

Reply:
Done with 41ebdfd69f3a1cda5e0fcfdc02693d615eba8d73. I will squash this commit into the original swap without scratch once approved.

Reply:
^^ patch looks good to me